File: go/internal/recovery/semanticoutcomes_test.go
package recovery

import (
	"testing"

	"github.com/johnny4young/janusly/go/internal/domain"
)

func v2Contract() *domain.RecoveryContract {
	contract := &domain.RecoveryContract{Version: "2"}
	contract.Failure.Semantic.Mode = "deterministic"
	contract.Failure.Semantic.Detectors = []domain.RecoverySemanticDetector{
		{
			ID: "det-total", SourceNodeID: "calc", Kind: "expression",
			PassWhen: "context.calc.output.total <= 100",
			Action:   "quarantine", Message: "total exceeds cap",
		},
		{
			ID: "det-shape", SourceNodeID: "calc", Kind: "schema",
			Schema: &domain.InputSchema{Type: "object",
				Properties: map[string]*domain.InputSchema{"total": {Type: "number"}},
				Required:   []string{"total"}},
			Action: "observe", Message: "output shape drifted",
		},
	}
	contract.Failure.Semantic.EvaluationFixtures = []domain.RecoverySemanticFixture{
		{ID: "fx-pass", SourceNodeID: "calc", Output: map[string]any{"total": float64(50)}, Expected: "pass"},
		{ID: "fx-violation", SourceNodeID: "calc", Output: map[string]any{"total": float64(500)}, Expected: "violation"},
	}
	return contract
}

func evaluate(contract *domain.RecoveryContract, node string, output any, context map[string]any) SemanticOutcomeEvaluation {
	return EvaluateSemanticOutcome(struct {
		Contract     *domain.RecoveryContract
		SourceNodeID string
		Output       any
		Context      map[string]any
	}{Contract: contract, SourceNodeID: node, Output: output, Context: context})
}

// The deterministic evaluator: expression + schema detectors over the
// overlaid completed output; quarantine dominates observe for the verdict;
// an expression ERROR counts as a violation with details (fail-safe).
func TestEvaluateSemanticOutcome(t *testing.T) {
	contract := v2Contract()

	pass := evaluate(contract, "calc", map[string]any{"total": float64(50)}, nil)
	if pass.Evaluated != 2 || len(pass.Violations) != 0 || pass.Quarantined {
		t.Fatalf("passing output: %+v", pass)
	}

	// Expression violation (quarantine) + schema pass.
	over := evaluate(contract, "calc", map[string]any{"total": float64(500)}, nil)
	if len(over.Violations) != 1 || !over.Quarantined || over.Violations[0].DetectorID != "det-total" {
		t.Fatalf("expression violation: %+v", over)
	}

	// Schema violation alone (observe) must NOT quarantine.
	shape := evaluate(contract, "calc", map[string]any{"total": "mucho"}, nil)
	quarantineSeen := false
	observeSeen := false
	for _, violation := range shape.Violations {
		if violation.DetectorID == "det-shape" && violation.Action == "observe" {
			observeSeen = true
		}
		if violation.Action == "quarantine" {
			quarantineSeen = true
		}
	}
	if !observeSeen {
		t.Fatalf("schema detector must fire: %+v", shape.Violations)
	}
	// Same-source verdict: with BOTH firing (string breaks the expression
	// comparison too), quarantine dominates — the strictest action governs.
	if quarantineSeen && !shape.Quarantined {
		t.Fatalf("quarantine must dominate: %+v", shape)
	}

	// A broken expression is a violation WITH details, never a silent pass.
	broken := v2Contract()
	broken.Failure.Semantic.Detectors = broken.Failure.Semantic.Detectors[:1]
	broken.Failure.Semantic.Detectors[0].PassWhen = "context.calc.output.total &&& nonsense"
	verdict := evaluate(broken, "calc", map[string]any{"total": float64(10)}, nil)
	if len(verdict.Violations) != 1 || len(verdict.Violations[0].Details) == 0 {
		t.Fatalf("broken expression must violate with details: %+v", verdict)
	}

	// A V1 contract NEVER evaluates: zero detectors by construction.
	v1 := &domain.RecoveryContract{Version: "1"}
	v1.Failure.Semantic.Mode = "disabled"
	if got := evaluate(v1, "calc", map[string]any{"total": float64(999)}, nil); got.Evaluated != 0 || len(got.Violations) != 0 {
		t.Fatalf("v1 must never evaluate: %+v", got)
	}

	// Detectors on OTHER nodes stay out of this node's verdict.
	other := evaluate(contract, "otro", map[string]any{"total": float64(999)}, nil)
	if other.Evaluated != 0 {
		t.Fatalf("foreign-node detectors must not run: %+v", other)
	}

	// Cross-node context flows into expressions.
	cross := v2Contract()
	cross.Failure.Semantic.Detectors = []domain.RecoverySemanticDetector{{
		ID: "det-cross", SourceNodeID: "calc", Kind: "expression",
		PassWhen: "context.calc.output.total <= context.limits.output.cap",
		Action:   "quarantine", Message: "over the configured cap",
	}}
	verdictCross := evaluate(cross, "calc", map[string]any{"total": float64(80)},
		map[string]any{"limits": map[string]any{"output": map[string]any{"cap": float64(100)}}})
	if len(verdictCross.Violations) != 0 {
		t.Fatalf("cross-node pass: %+v", verdictCross.Violations)
	}
}

// Fixture replay uses the exact runtime evaluator, verdicts included.
func TestEvaluateSemanticOutcomeFixtures(t *testing.T) {
	results := EvaluateSemanticOutcomeFixtures(v2Contract())
	if len(results) != 2 {
		t.Fatalf("two fixtures: %+v", results)
	}
	for _, result := range results {
		if !result.Passed {
			t.Fatalf("fixture %s must match its expectation: %+v", result.ID, result)
		}
	}
	// V1 → no fixtures ever.
	v1 := &domain.RecoveryContract{Version: "1"}
	if got := EvaluateSemanticOutcomeFixtures(v1); len(got) != 0 {
		t.Fatalf("v1 fixtures must be empty: %+v", got)
	}
}

File: packages/engine/src/semantic-outcomes.test.ts
import { describe, expect, it } from "vitest";
import type { RecoveryContractV2 } from "@janusly/shared";
import {
  evaluateSemanticOutcome,
  evaluateSemanticOutcomeFixtures,
} from "./semantic-outcomes";

function contract(): RecoveryContractV2 {
  return {
    version: "2",
    failure: {
      technical: {
        terminalNodeFailure: true,
        stalledNode: true,
      },
      semantic: {
        mode: "deterministic",
        detectors: [
          {
            id: "mode",
            sourceNodeId: "answer",
            kind: "expression",
            passWhen: 'context.answer.output.mode === "ai"',
            action: "quarantine",
            message: "AI mode is required",
          },
          {
            id: "shape",
            sourceNodeId: "answer",
            kind: "schema",
            schema: {
              type: "object",
              required: ["text"],
              properties: {
                text: { type: "string" },
              },
            },
            action: "observe",
            message: "Text is required",
          },
        ],
        evaluationFixtures: [
          {
            id: "good",
            sourceNodeId: "answer",
            output: { mode: "ai", text: "ready" },
            expected: "pass",
          },
          {
            id: "bad",
            sourceNodeId: "answer",
            output: { mode: "fallback" },
            expected: "violation",
          },
        ],
      },
    },
    evidence: {
      required: [
        "failure_snapshot",
        "audit_trail",
        "validation_receipt",
        "terminal_outcome",
      ],
    },
    effects: [],
    repairs: { allowed: ["retry"] },
    validation: { minimumEvidenceLevel: "writes_skipped" },
    approval: {
      productionMutation: "required",
      permission: "recovery.write",
    },
    autonomyLevel: 3,
    verification: {
      kind: "generation_bound_terminal_success",
    },
    recurrence: { windowDays: 7 },
  };
}

describe("semantic outcome evaluator", () => {
  it("overlays the just-completed output and quarantines only on a failed quarantine detector", () => {
    const good = evaluateSemanticOutcome({
      contract: contract(),
      sourceNodeId: "answer",
      output: { mode: "ai", text: "ready" },
      context: {
        answer: { status: "running", output: {} },
      },
    });
    expect(good).toEqual({
      sourceNodeId: "answer",
      evaluated: 2,
      violations: [],
      quarantined: false,
    });

    const bad = evaluateSemanticOutcome({
      contract: contract(),
      sourceNodeId: "answer",
      output: { mode: "fallback" },
    });
    expect(bad.quarantined).toBe(true);
    expect(bad.violations.map((item) => item.detectorId)).toEqual([
      "mode",
      "shape",
    ]);
  });

  it("replays bounded fixtures through the runtime evaluator", () => {
    expect(evaluateSemanticOutcomeFixtures(contract())).toEqual([
      expect.objectContaining({
        id: "good",
        expected: "pass",
        actual: "pass",
        passed: true,
      }),
      expect.objectContaining({
        id: "bad",
        expected: "violation",
        actual: "violation",
        passed: true,
      }),
    ]);
  });

  it("evaluates an explicit null output without coercing it to an object", () => {
    const nullContract: RecoveryContractV2 = {
      ...contract(),
      failure: {
        ...contract().failure,
        semantic: {
          mode: "deterministic",
          detectors: [
            {
              id: "nullable",
              sourceNodeId: "answer",
              kind: "expression",
              passWhen: "context.answer.output === null",
              action: "quarantine",
              message: "Null is required",
            },
          ],
          evaluationFixtures: [
            {
              id: "null-pass",
              sourceNodeId: "answer",
              output: null,
              expected: "pass",
            },
            {
              id: "object-fail",
              sourceNodeId: "answer",
              output: {},
              expected: "violation",
            },
          ],
        },
      },
    };

    expect(
      evaluateSemanticOutcome({
        contract: nullContract,
        sourceNodeId: "answer",
        output: null,
      }),
    ).toMatchObject({ violations: [], quarantined: false });
  });

  it("leaves v1 contracts and unrelated nodes untouched", () => {
    expect(
      evaluateSemanticOutcome({
        sourceNodeId: "other",
        output: {},
        contract: contract(),
      }),
    ).toEqual({
      sourceNodeId: "other",
      evaluated: 0,
      violations: [],
      quarantined: false,
    });
  });
});
